modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -79,43 +108,125 @@ object GlobalL0Service {
           } { lastStoredOrdinal =>
-            def pulled = globalL0ClusterStorage.getRandomPeer.flatMap { l0Peer =>
-              l0GlobalSnapshotClient.getLatestOrdinal
-                .run(l0Peer)
-                .map { lastOrdinal =>
-                  val nextOrdinal = lastStoredOrdinal.next
-                  val lastOrdinalCap = lastOrdinal.value.value
-                    .min(singlePullLimit.map(nextOrdinal.value.value + _.value).getOrElse(lastOrdinal.value.value))
-
-                  nextOrdinal.value.value to lastOrdinalCap
-                }
-                .map(_.toList.map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o))))
-                .flatMap { ordinals =>
-                  (ordinals, List.empty[Hashed[GlobalIncrementalSnapshot]]).tailRecM {
-                    case (ordinal :: nextOrdinals, snapshots) =>
-                      l0GlobalSnapshotClient
-                        .get(ordinal)(l0Peer)
-                        .flatMap(_.toHashedWithSignatureCheck.flatMap(_.liftTo[F]))
-                        .map(s => (nextOrdinals, snapshots :+ s).asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
-                        .handleErrorWith { e =>
-                          logger
-                            .warn(e)(s"Failure pulling snapshot with ordinal=$ordinal")
-                            .map(_ => snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
-                        }
-
-                    case (Nil, snapshots) =>
-                      Applicative[F].pure(snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
-                  }
-                }
-            }
-
-            pulled.map(_.asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)])
+            for {
+              l0Peer <- globalL0ClusterStorage.getRandomPeer
+              lastOrdinal <- l0GlobalSnapshotClient.getLatestOrdinal.run(l0Peer)
+              pulled <- pullSnapshots(l0Peer, lastStoredOrdinal.next, lastOrdinal)
+            } yield pulled.asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)]
           }
         }.handleErrorWith { e =>
           logger.warn(e)(s"Failure pulling global snapshots!") >>
             Applicative[F].pure(
               List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)]
             )
         }
+
+      def pullGlobalSnapshotsFromMajority
+        : F[Either[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo), List[Hashed[GlobalIncrementalSnapshot]]]] = {
+        maybeMajorityPeerIds match {
+          case None => pullGlobalSnapshots
+          case Some(majorityPeerIds) =>
+            lastGlobalSnapshotStorage.getOrdinal.flatMap {
+              _.fold {
+                pullLatestSnapshotFromMajority.map(_.asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+              } { lastStoredOrdinal =>
```

Comment:
Why you use:
```scala
maybe... match {
 case None => ...
 case Some(_) => ...
}
```

and few lines later you use:
```scala
maybe.fold { ... } { _ => ... }
```
?

I'd keep using the same pattern across the file (I prefer fold, but whatever works for you better).

Reply:
changed